build/builder.go

```diff
@@ -942,6 +945,40 @@ func sortDocuments(todo []*zoekt.Document) {
 	}
 }
 
+// sortDocuments2 sorts zoekt.Document according to their Scores. In general,
+// documents can have a nil score vector if the document to be indexed was
+// introduced after the ranking took place. A nil score vector translates to the
+// lowest possible rank.
+func sortDocuments2(rs []*zoekt.Document) {
+	sort.Slice(rs, func(i, j int) bool {
+		r1 := rs[i].Scores
+		r2 := rs[j].Scores
+
+		if r1 == nil && r2 == nil {
+			return false
+		}
+
+		if r1 == nil {
+			return false
+		}
+
+		if r2 == nil {
+			return true
+		}
+
+		for i := range r1 {
+			if r1[i] < r2[i] {
+				return true
+			}
+			if r1[i] > r2[i] {
+				return false
+			}
+		}
+
+		return false
```

Comment:
I ended up writing an alternative implementation which handled different lengths just in case that happened (and to ensure I understood this logic). But I suppose it is quite terse and this implementation may be better. Feel free to decide :)

```suggestion
		l := len(r1)
		if len(r2) < l {
		  l = len(r2)
		}
		for i := 0; i < l; i++ {
		  if r1[i] != r2[i] {
		    return r1[i] < r2[i]
		  }
		}
		// if r2 has more scores it is more important. ie imagine right padding arrays with zeros so they are the same length.
		return len(r1) < len(r2)
```